File: spark/src/main/scala/geotrellis/spark/io/AttributeStore.scala
/*
 * Copyright 2016 Azavea
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package geotrellis.spark.io

import geotrellis.spark._
import geotrellis.spark.io.index._
import geotrellis.spark.io.json.Implicits._

import org.apache.avro.Schema
import spray.json._
import spray.json.DefaultJsonProtocol.JsValueFormat

import scala.reflect._
import java.net.URI
import java.util.ServiceLoader

trait AttributeStore extends  AttributeCaching with LayerAttributeStore {
  def read[T: JsonFormat](layerId: LayerId, attributeName: String): T
  def readAll[T: JsonFormat](attributeName: String): Map[LayerId, T]
  def write[T: JsonFormat](layerId: LayerId, attributeName: String, value: T): Unit
  def layerExists(layerId: LayerId): Boolean
  def delete(layerId: LayerId): Unit
  def delete(layerId: LayerId, attributeName: String): Unit
  def layerIds: Seq[LayerId]
  def availableAttributes(id: LayerId): Seq[String]

  def copy(from: LayerId, to: LayerId): Unit =
    copy(from, to, availableAttributes(from))

  def copy(from: LayerId, to: LayerId, attributes: Seq[String]): Unit = {
    for (attribute <- attributes) {
      write(to, attribute, read[JsValue](from, attribute))
    }
  }
}

object AttributeStore {
  // TODO: Move to LayerAttributeStore, and potentially rename that to
  // AvroLayerAttributeStore (these things are not needed for COGs)
  object Fields {
    val metadataBlob = "metadata"
    val header = "header"
    val keyIndex = "keyIndex"
    val metadata = "metadata"
    val schema = "schema"
  }

  /**
   * Produce AttributeStore instance based on URI description.
   * This method uses instances of [[AttributeServiceProvider]] loaded through Java SPI.
   */
  def apply(uri: URI): AttributeStore = {
    import scala.collection.JavaConversions._

    ServiceLoader.load(classOf[AttributeStoreProvider]).iterator()
      .find(_.canProcess(uri))
      .getOrElse(throw new RuntimeException(s"Unable to find AttributeStoreProvider for $uri"))
      .attributeStore(uri)
  }

  def apply(uri: String): AttributeStore = apply(new URI(uri))
}

case class LayerAttributes[H, M, K](header: H, metadata: M, keyIndex: KeyIndex[K], schema: Schema)

trait LayerAttributeStore extends Serializable {
  def readHeader[H: JsonFormat](id: LayerId): H
  def readMetadata[M: JsonFormat](id: LayerId): M
  def readKeyIndex[K: ClassTag](id: LayerId): KeyIndex[K]
  def readSchema(id: LayerId): Schema
  def readLayerAttributes[H: JsonFormat, M: JsonFormat, K: ClassTag](id: LayerId): LayerAttributes[H, M, K]
  def writeLayerAttributes[H: JsonFormat, M: JsonFormat, K: ClassTag](id: LayerId, header: H, metadata: M, keyIndex: KeyIndex[K], schema: Schema): Unit
}


trait BlobLayerAttributeStore extends AttributeStore {
  import AttributeStore._
  import DefaultJsonProtocol._

  def readHeader[H: JsonFormat](id: LayerId): H =
    cacheRead[JsValue](id, Fields.metadataBlob).asJsObject.fields(Fields.header).convertTo[H]

  def readMetadata[M: JsonFormat](id: LayerId): M =
    cacheRead[JsValue](id, Fields.metadataBlob).asJsObject.fields(Fields.metadata).convertTo[M]

  def readKeyIndex[K: ClassTag](id: LayerId): KeyIndex[K] =
    cacheRead[JsValue](id, Fields.metadataBlob).asJsObject.fields(Fields.keyIndex).convertTo[KeyIndex[K]]

  def readSchema(id: LayerId): Schema =
    cacheRead[JsValue](id, Fields.metadataBlob).asJsObject.fields(Fields.schema).convertTo[Schema]

  def readLayerAttributes[H: JsonFormat, M: JsonFormat, K: ClassTag](id: LayerId): LayerAttributes[H, M, K] = {
    val blob = cacheRead[JsValue](id, Fields.metadataBlob).asJsObject
    LayerAttributes(
      blob.fields(Fields.header).convertTo[H],
      blob.fields(Fields.metadata).convertTo[M],
      blob.fields(Fields.keyIndex).convertTo[KeyIndex[K]],
      blob.fields(Fields.schema).convertTo[Schema]
    )
  }

  def writeLayerAttributes[H: JsonFormat, M: JsonFormat, K: ClassTag](id: LayerId, header: H, metadata: M, keyIndex: KeyIndex[K], schema: Schema): Unit = {
    cacheWrite(id, Fields.metadataBlob,
      JsObject(
        Fields.header -> header.toJson,
        Fields.metadata -> metadata.toJson,
        Fields.keyIndex -> keyIndex.toJson,
        Fields.schema -> schema.toJson
      )
    )
  }
}

trait DiscreteLayerAttributeStore extends AttributeStore {
  import AttributeStore._

  def readHeader[H: JsonFormat](id: LayerId): H =
    cacheRead[H](id, Fields.header)

  def readMetadata[M: JsonFormat](id: LayerId): M =
    cacheRead[M](id, Fields.metadata)

  def readKeyIndex[K: ClassTag](id: LayerId): KeyIndex[K] =
    cacheRead[KeyIndex[K]](id, Fields.keyIndex)

  def readSchema(id: LayerId): Schema =
    cacheRead[Schema](id, Fields.schema)

  def readLayerAttributes[H: JsonFormat, M: JsonFormat, K: ClassTag](id: LayerId): LayerAttributes[H, M, K] = {
    LayerAttributes(
      readHeader[H](id),
      readMetadata[M](id),
      readKeyIndex[K](id),
      readSchema(id)
    )
  }

  def writeLayerAttributes[H: JsonFormat, M: JsonFormat, K: ClassTag](id: LayerId, header: H, metadata: M, keyIndex: KeyIndex[K], schema: Schema) = {
    cacheWrite(id, Fields.header, header)
    cacheWrite(id, Fields.metadata, metadata)
    cacheWrite(id, Fields.keyIndex, keyIndex)
    cacheWrite(id, Fields.schema, schema)
  }
}

File: geotrellis-python/geotrellis/spark/io/AttributeStore.py
from __future__ import absolute_import
from __future__ import absolute_import
from geotrellis.spark.io.index.KeyIndex import KeyIndex
from geotrellis.spark.io.AttributeCaching import AttributeCaching
from geotrellis.python.util.utils import get_format_for_key_index_type, SchemaFormat
import avro.schema
import json

class AttributeStore(AttributeCaching):
    def __init__(self):
        super(AttributeStore, self).__init__()
    def availableAttributes(self, layer_id):
        pass

    def copy(self, from_layer_id, to_layer_id, attrs = None):
        if attrs is None:
            attrs = self.availableAttributes(from_layer_id)
        for attr_name in attrs:
            self.write(to_layer_id, attr_name, self.read(from_layer_id, attr_name))

class Fields(object):
    metadataBlob = "metadata"
    header = "header"
    keyIndex = "keyIndex"
    metadata = "metadata"
    schema = "schema"

class LayerAttributes(object):
    def __init__(self, header, metadata, keyIndex, schema):
        self.header = header
        self.metadata = metadata
        self.keyIndex = keyIndex
        self.schema = schema

class BlobLayerAttributeStore(AttributeStore):
    def __init__(self):
        super(BlobLayerAttributeStore, self).__init__()
    def readHeader(self, header_type, layer_id):
        header = self.cacheRead(dict, layer_id, Fields.metadataBlob)[Fields.header]
        header_format = header_type.implicits['format']()
        return header_format.from_dict(header)

    def readMetadata(self, metadata_type, layer_id):
        meta = self.cacheRead(dict, layer_id, Fields.metadataBlob)[Fields.metadata]
        metadata_format = metadata_type.implicits['format']()
        return metadata_format.from_dict(meta)

    def readKeyIndex(self, key_type, layer_id):
        key_index = self.cacheRead(dict, layer_id, Fields.metadataBlob)[Fields.keyIndex]
        # TODO temporary workaround. see geotrellis.spark.io.json.KeyIndexFormats
        #key_index_format = get_format_for_key_index_type(KeyIndex[key_type])
        key_index_format = KeyIndex[key_type].implicits['format']()
        return key_index_format.from_dict(key_index)

    def readSchema(self, layer_id):
        schema_as_dict = self.cacheRead(dict, layer_id, Fields.metadataBlob)[Fields.schema]
        schema_json_string = json.dumps(schema_as_dict)
        return avro.schema.parse(schema_json_string)

    def readLayerAttributes(self, header_type, metadata_type, key_type, layer_id):
        headerFormat =    header_type.implicits['format']()
        metaFormat =    metadata_type.implicits['format']()
        keyindFormat = KeyIndex[key_type].implicits['format']()
        schemaFormat = SchemaFormat()
        blob = self.cacheRead(dict, layer_id, Fields.metadataBlob)
        return LayerAttributes(
                    headerFormat.from_dict(blob[Fields.header]),
                    metaFormat.from_dict(blob[Fields.metadata]),
                    keyindFormat.from_dict(blob[Fields.keyIndex]),
                    schemaFormat.from_dict(blob[Fields.schema]))

    def writeLayerAttributes(self, header_type, metadata_type, key_index_type, layer_id, header, meta, key_index, schema):
        headerFormat =    header_type.implicits['format']()
        metaFormat =    metadata_type.implicits['format']()
        keyindFormat = key_index_type.implicits['format']()
        schemaFormat = SchemaFormat()
        jsobj = {
                Fields.header:      headerFormat.to_dict(header),
                Fields.metadata:    metaFormat.to_dict(meta),
                Fields.keyIndex:    keyindFormat.to_dict(key_index),
                Fields.schema:      schemaFormat.to_dict(schema)
            }
        self.cacheWrite(dict, layer_id, Fields.metadataBlob, jsobj)

